### PR TITLE
Issue 4 - Implementation of cosmoz-tree logic

### PR DESCRIPTION
**Pull request by JaySunSyn**
Needs to be synced with https://github.com/Neovici/cosmoz-tree/pull/1

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -20,7 +20,8 @@
     "iron-flex-layout": "^1.0.0",
     "cosmoz-i18next": "neovici/cosmoz-i18next#master",
     "cosmoz-dialog": "neovici/cosmoz-dialog#master",
-    "iron-list": "PolymerElements/iron-list#^2.0.4"
+    "iron-list": "PolymerElements/iron-list#^2.0.4",
+    "cosmoz-treenode": "Neovici/cosmoz-treenode"
```
Comment (nomego) [inline bower.json]: why was this added?

File: cosmoz-treenode-navigator.js [hunk 2]
```diff
@@ -282,35 +266,22 @@
...
-					return node;
-				}
-				if (node[childProperty] !== undefined) {
-					var children = node[childProperty];
-					children = Object.keys(children).map(function (key) { return children[key]; });
-					this._findInNodes(results, children, searchStr, searchAttr, childProperty, data);
-				}
-			}, this)
+			var nodes = {};
+			nodeList.forEach(function (n) { nodes[n.id] = n; });
```
Comment (nomego) [inline cosmoz-treenode-navigator.js]: this should be a `reduce()`

File: cosmoz-treenode-navigator.js [hunk 3]
```diff
@@ -282,35 +266,22 @@
...
+
+			var tree = new Cosmoz.DefaultTree(nodes);
```
Comment (nomego) [inline cosmoz-treenode-navigator.js]: `this.tree` should be supplied instead of `this.data`

File: cosmoz-treenode-navigator.js [hunk 4]
```diff
@@ -282,35 +266,22 @@
...
+			
+			tree.getPath = function (node) {
```
Comment (nomego) [inline cosmoz-treenode-navigator.js]: why is this method created?

File: bower.json (hunk 1 above)
Comment (nomego) [inline bower.json]: No response to this.
I think this is an unnecessary dep/import.

File: cosmoz-treenode-navigator.html [hunk 5]
```diff
@@ -62,7 +64,7 @@ <h3 class="layout horizontal center wrap">
 					<div class="section" hidden$="[[ !_renderSection(_searching, index, dataPlane, node) ]]">[[ node.sectionName ]]</div>
 					<paper-item on-tap="_nodeSelected" class="nodeItem">
 						<span class="flex">[[ node.name ]]</span>
-						<iron-icon hidden$="[[ !hasChildren(node) ]]" icon="icons:arrow-forward" on-click="openNode" data-path$="[[ node.path ]]"></iron-icon>
+						<iron-icon hidden$="[[ !hasChildren(node) ]]" data-path$="[[ node.path ]]" icon="icons:arrow-forward" on-click="openNode" data-path$="[[ node.path ]]"></iron-icon>
```
Comment (nomego) [inline cosmoz-treenode-navigator.html]: `data-path` was already bound

File: cosmoz-treenode-navigator.js [hunk 6]
```diff
@@ -15,24 +16,26 @@
 			that component is given
 			 */
 			data: {
```
Comment (nomego) [inline cosmoz-treenode-navigator.js]: I don't think `data` should be kept but rather just expect a
```js
tree: {
  type: Cosmoz.Tree
}
```

like https://github.com/Neovici/cosmoz-treenode/blob/master/cosmoz-treenode.html#L45

File: cosmoz-treenode-navigator.js [hunk 7]
```diff
@@ -15,24 +16,26 @@
...
-				type: Object
+				type: Object,
+				value: function () { return {} },
```
Comment (nomego) [inline cosmoz-treenode-navigator.js]: `return` on separate line, missing semicolon

File: cosmoz-treenode-navigator.js [hunk 8]
```diff
@@ -162,108 +178,38 @@
...
-		_renderLevel: function (pl, nodes) {
-			var pathSegment = nodes,
-				pathArray,
-				level = [];
-			if (!pl) {
-				// Return the formatted root nodes.
-				level = Object.keys(nodes).map(function (key) {
-					var node = nodes[key];
+		_renderLevel: function (pathLocator, tree) {
+			if (!tree) return;
```
Comment (nomego) [inline cosmoz-treenode-navigator.js]: use curly braces

File: cosmoz-treenode-navigator.js [hunk 9]
```diff
@@ -162,108 +178,38 @@
...
-					return true;
-				}
-				var children = node[this.childProperty];
-				if (children !== undefined && Object.keys(children).length > 0) {
-					pathSegment = children;
-				}
-			}, this);
-
-			return node;
+		_getNode: function (pathLocator, nodes) {
```
Comment (nomego) [inline cosmoz-treenode-navigator.js]: second argument should be dropped if not used (also from computed property)

File: cosmoz-treenode-navigator.js [hunk 10]
```diff
@@ -162,108 +178,38 @@
...
+			if (!pathLocator) return null;
```
Comment (nomego) [inline cosmoz-treenode-navigator.js]: use curly braces

File: cosmoz-treenode-navigator.js [hunk 11]
```diff
@@ -162,108 +178,38 @@
...
-				if (!node) {
-					return true;
-				}
-				node.path = newPathString;
-				parts.push(node);
-			}, this);
-
-			return parts;
+		_getTreePathParts: function (pathLocator, tree) {
+			if (!tree) return;
```
Comment (nomego) [inline cosmoz-treenode-navigator.js]: use curly braces (x2), why different return values depending on missing argument?

File: cosmoz-treenode-navigator.js [hunk 12]
```diff
@@ -281,53 +227,12 @@
...
-				var path = element.dataset.path;
-				if (path !== undefined) {
-					this._openNodeLevelPath = path;
-					return true;
-				} else if (index > 3) {
-					return true;
-				}
-			}, this);
-
+			this._openNodeLevelPath = event.currentTarget.getAttribute('data-path');
```
Comment (nomego) [inline cosmoz-treenode-navigator.js]: why not `event.currentTarget.dataset.path` ?

File: cosmoz-treenode-navigator.js [hunk 13]
```diff
@@ -141,13 +144,26 @@
...
+				return tree.getNodeByProperty(this.comparisonProperty, inputValue, false, true, renderedLevel);
 			}
 			return renderedLevel;
 		},
 		/**
+		 * A hook to manipulate the default tree class
+		 * of Cosmoz.TreeBehavior
+		 */
+		_getTreeClass: function (treeClass) {
+			treeClass.prototype.afterFound = function (node, propertyName) {
```
Comment (nomego) [inline cosmoz-treenode-navigator.js]: Any reason why we can't just do the equivalent in the `map()` of `_renderLevel` ?
This code both updates an unknown tree implementation (once `data` is replaced with `tree`) and it writes to the nodes in the tree.

File: bower.json (hunk 1 above)
Comment (JaySunSyn) [inline bower.json]: @nomego  Ok, not sure I get this and I think its related with the behavior I created. 

Do I understand correctly, `cosmoz-treenode-navigator` gets a tree from outside the element? If so, how can I be sure that certain methods of `Cosmoz.DefaultTree` are implemented in the tree data the navigator gets from outside in order to work with them inside of the navigator? Do we define, the data (tree object) we get from outside needs do be a `Cosmoz.DefaultTree` (or derived from) object?

My understanding was, we get "node-like" data from outside and use the Cosmoz.Tree or DefaultTree methods inside the navigator. => wrong?

Comment (JaySunSyn) [inline bower.json]: Ok I see the mistake now, cosmoz-tree instead of cosmoz-treenode. Sorry for that, but above question still valid.

File: cosmoz-treenode-navigator.js (hunk 6 above)
Comment (JaySunSyn) [inline cosmoz-treenode-navigator.js]: I guess that answers a big part of my above question.

File: bower.json (hunk 1 above)
Comment (nomego) [inline bower.json]: Yes, the navigator gets a `Cosmoz.Tree` from outside the element.
It can be a `Cosmoz.DefaultTree` or some other implementation.
`Cosmoz.Tree` acts as an "abstract class" and defines methods that the implementations need to implement.

Yes the last part is wrong, we expect the tree to be created for us.

Also, it's a weird place to comment about this. This change is to import the "cosmoz-treenode" element which we haven't really touched during this PR, so I don't see why the dependency was added.
Ah, now I realize! :) It's just a typo, it should depend on `Neovici/cosmoz-tree#master` rather than `Neovici/cosmoz-treenode` :)

File: cosmoz-treenode-navigator.js (hunk 13 above)
Comment (JaySunSyn) [inline cosmoz-treenode-navigator.js]: We can. (just a tiny little bit less efficient but no big deal)
 I'm just used to such hook methods I know from js-data.

File: bower.json [hunk 14]
```diff
@@ -26,6 +26,11 @@
     "iron-component-page": "PolymerElements/iron-component-page#^1.0.0",
     "iron-demo-helpers": "PolymerElements/iron-demo-helpers#^1.0.0",
     "web-component-tester": "^4.0.0",
-    "webcomponentsjs": "webcomponents/webcomponentsjs#^0.7.0"
+    "webcomponentsjs": "webcomponents/webcomponentsjs#^0.7.0",
+    "cosmoz-tree": "Neovici/cosmoz-tree",
+    "iron-ajax": "^2.0.2"
+  },
+  "resolutions": {
+    "iron-ajax": "^1.0.0"
```
Comment (nomego) [inline bower.json]: Should not be needed if the dep is on iron-ajax#^1.0.0

File: bower.json [hunk 15]
```diff
@@ -26,6 +26,11 @@
     "iron-component-page": "PolymerElements/iron-component-page#^1.0.0",
     "iron-demo-helpers": "PolymerElements/iron-demo-helpers#^1.0.0",
     "web-component-tester": "^4.0.0",
-    "webcomponentsjs": "webcomponents/webcomponentsjs#^0.7.0"
+    "webcomponentsjs": "webcomponents/webcomponentsjs#^0.7.0",
+    "cosmoz-tree": "Neovici/cosmoz-tree",
+    "iron-ajax": "^2.0.2"
```
Comment (nomego) [inline bower.json]: ^1.0.0

File: cosmoz-treenode-navigator.js [hunk 16]
```diff
@@ -5,7 +5,7 @@
 	Polymer({
 
 		behaviors: [
-			Cosmoz.TranslatableBehavior
+			Cosmoz.TranslatableBehavior,
```
Comment (nomego) [inline cosmoz-treenode-navigator.js]: drop comma

File: cosmoz-treenode-navigator.js [hunk 17]
```diff
@@ -162,108 +163,49 @@
...
-				
-			return path.map(function (nodeKey) {
-				var node = pathSegment[nodeKey],
-					children = node[this.childProperty];
-				node['key'] = nodeKey;
-				if (node && children !== undefined && Object.keys(children).length > 0) {
-					pathSegment = children;
-				}
-				return node;
+		_normalizeNode: function (nodeObj) {
```
Comment (nomego) [inline cosmoz-treenode-navigator.js]: This method actually doesn't return a formatted node as described and expected by the name but rather an array based on the object keys. It is designed to work on `node.children` input, but is not used that way.
Keep only the object remapping (1:1, return statement) in this method.
Create a new method for `_normalizeNodes()` that takes an array as input (and maps it by calling _normalizeNode() for each item). Probably needed when doing searches?
Finally, use `_normalizeNodes(tree.getChildren(node))` or even a `_normalizeChildren()` for cases where you want this code to run as it is now.

File: cosmoz-treenode-navigator.js [hunk 18]
```diff
@@ -162,108 +163,49 @@
...
-				node.path = newPathString;
-				parts.push(node);
-			}, this);
-
-			return parts;
+		_getTreePathParts: function (pathLocator, tree) {
+			if (!tree || !pathLocator) {
+				return [];
+			};
+			return this._normalizeNode(tree.getPath(pathLocator));
```
Comment (nomego) [inline cosmoz-treenode-navigator.js]: `getPath()` returns an array, need `_normalizeNodes()`
But I don't think we need to normalize these, do we ?

File: cosmoz-treenode-navigator.js [hunk 19]
```diff
@@ -162,108 +163,44 @@
...
-				node.path = newPathString;
-				parts.push(node);
-			}, this);
-
-			return parts;
+		_getTreePathParts: function (pathLocator, tree) {
+			if (!tree || !pathLocator) {
+				return [];
+			};
+			return this._normalizeNodes(tree.getPath(pathLocator));
```
Comment (nomego) [inline cosmoz-treenode-navigator.js]: Why do the nodes in the path need to be normalized?
Can't we refactor away `_getTreePathParts()` completely?

File: demo/index.html [hunk 20]
```diff
@@ -29,12 +25,21 @@
...
+					<iron-ajax url="data/multiRootTree.json" handle-as="json" on-response="handleResponse" debounce-duration="10" auto></iron-ajax>
+					<cosmoz-treenode-button-view id="navigator" value="{{ path }}" chosen-node="{{ nodeChosen1 }}"></cosmoz-treenode-button-view>
 				</template>
-			</demo-snippet>
-		</div>
+			  </demo-snippet>
+		 </div>
+		<script>
+			window.addEventListener('WebComponentsReady', function(e) {
+				document.querySelector('#demo').handleResponse = function(response) {
```
Comment (nomego) [inline demo/index.html]: This could actually be further simplified..
```html
<iron-ajax ... last-response="{{ treeData }}">
<cosmoz-treenode-button-view ... tree="[[ getTree(treeData) ]]">
```
```js
document.querySelector('#demo').getTree = function (data) {
  return new Cosmoz.DefaultTree(data);
};
```

Still, I don't think you need `window.addEventListener('WebComponentsReady', function(e) {` either?

File: cosmoz-treenode-navigator.js [hunk 21]
```diff
@@ -141,9 +141,10 @@
 				value: 1
 			}
 		},
-		_computeDataPlane: function (searching, inputValue, renderedLevel, openNodeLevelPathParts, data) {
+		_computeDataPlane: function (searching, inputValue, renderedLevel, openNodeLevelPathParts, tree) {
 			if (searching) {
-				return this.searchAllBranches(inputValue, openNodeLevelPathParts, renderedLevel, data);
+				var results = tree.getNodeByProperty(this.comparisonProperty, inputValue, false, true, renderedLevel);
```
Comment (JaySunSyn) [inline cosmoz-treenode-navigator.js]: weird.. I have this on my local copy..

File: cosmoz-treenode-navigator.js [hunk 22]
```diff
@@ -162,108 +163,44 @@
...
-				
-			return path.map(function (nodeKey) {
-				var node = pathSegment[nodeKey],
-					children = node[this.childProperty];
-				node['key'] = nodeKey;
-				if (node && children !== undefined && Object.keys(children).length > 0) {
-					pathSegment = children;
-				}
-				return node;
+		_normalizeNodes: function (nodes) {
```
Comment (JaySunSyn) [inline cosmoz-treenode-navigator.js]: I dropped the key property (which is getting lost when converting to object values) as it wasn't needed and now we can just call it on a resulting array.

File: demo/index.html (hunk 20 above)
Comment (JaySunSyn) [inline demo/index.html]: sweet!

File: cosmoz-treenode-navigator.js (hunk 19 above)
Comment (JaySunSyn) [inline cosmoz-treenode-navigator.js]: We need `node.path`in the path in order to open the node when clicked on a path part.

@nomego We could of course use `node.pathLocator` but it's nicer to have the same node structure in the view everywhere.

```
<span class="slash">/</span><a href="#" data-path$="[[ node.path ]]" on-click="openNode">[[ getNodeName(node) ]]</a>
```
But if you prefer I can do `data-path$="[[ node.pathLocator ]]"`?

Comment (nomego) [inline cosmoz-treenode-navigator.js]: Oh, ok, then I see.
Nah, keep it as it is now, it's better.

File: cosmoz-treenode-navigator.js [hunk 23]
```diff
@@ -162,108 +163,45 @@
...
-					}, this);
-				} else {
-					pathSegment = children;
-				}
-			}, this);
-
-			return this._sortItOut(level);
+			var node = tree.getNodeByPathLocator(pathLocator),
+				children = tree.getChildren(node),
+				level = children || _objectValues(node);
```
Comment (nomego) [inline cosmoz-treenode-navigator.js]: Why do we need `_objectValues()` on the node here?
Seems like it would render a weird result?

File: cosmoz-treenode-navigator.js [hunk 24]
```diff
@@ -162,108 +163,45 @@
...
-				node['key'] = nodeKey;
-				if (node && children !== undefined && Object.keys(children).length > 0) {
-					pathSegment = children;
-				}
-				return node;
+		_normalizeNodes: function (nodes) { 
+			return nodes.map(function (node) {
+				var path = node.pathLocator || node.path;
+				return {
+					name: node[this.comparisonProperty],
```
Comment (nomego) [inline cosmoz-treenode-navigator.js]: `this.comparisonProperty` should be replaced with `this.tree.searchProperty` wherever possible.
Same goes for `this.childProperty` -> `this.tree.childProperty`.

File: cosmoz-treenode-button-view.html [hunk 25]
```diff
@@ -52,12 +52,15 @@
...
 			<cosmoz-treenode-navigator id="treeNavigator" chosen-node="{{ chosenNode }}" tree="[[ tree ]]"
-				id="treenodeNavigator" on-data-plane-changed="refit"
-				highlighted-node-path="{{ highlightedNodePath }}" child-property="[[ childProperty ]]"
-				comparison-property="[[ comparisonProperty ]]" search-placeholder="[[ searchPlaceholder ]]"
-				separator-sign="[[ separatorSign ]]" local-search-done-text="[[ localSearchDoneText ]]"
-				reset-text="[[ resetText ]]" search-min-length="[[ searchMinLength ]]"
-				value="{{ value }}" value-path-parts="{{ valuePathParts }}">
+				on-data-plane-changed="refit"
+				highlighted-node-path="{{ highlightedNodePath }}"
+				comparison-property="[[ comparisonProperty ]]"
```
Comment (nomego) [inline cosmoz-treenode-button-view.html]: Drop comparisonProperty here as well

File: cosmoz-treenode-navigator.js [hunk 26]
```diff
@@ -67,7 +63,7 @@
 
 			chosenNode: {
 				type: Object,
-				computed: '_getNode(value, data)',
+				computed: '_getNode(value)',
```
Comment (plequang) [inline cosmoz-treenode-navigator.js]: The `_getNode` function needs to be bound to the tree, so that it will be called only when the tree is available.

File: cosmoz-treenode-navigator.js [hunk 27]
```diff
@@ -162,172 +135,67 @@
...
+				};
 			}, this);
 		},
 		/**
 		 * Returns a node based on a given path locator.
+		 * If pathLocator is empty or not defined, null gets returned.
 		 */
-		_getNode: function (pl, nodes) {
-			if (!pl) {
+		_getNode: function (pathLocator) {
```
Comment (plequang) [inline cosmoz-treenode-navigator.js]: This should be `_getNode: function (pathLocator, tree)`  and ensure `tree` is defined.
